### ISSUE TITLE
Proxy fails with "router could not recognize request"

### ISSUE THREAD

**Issue by timfpark** [linkerd/linkerd2#1861]
# Thanks for your help improving the project! #

## Getting Help ##

Github issues are for bug reports and feature requests. For questions about
Linkerd, how to use it, or debugging assistance, start by
[asking a question in the forums](https://discourse.linkerd.io/) or join us on
[Slack](https://slack.linkerd.io/).

Full details at [CONTRIBUTING.md](https://github.com/linkerd/linkerd/blob/master/CONTRIBUTING.md).

## Filing a Linkerd issue ##

Issue Type:

- [ X ] Bug report
- [ ] Feature request

**What happened**:

After injecting linkerd on my service using `linkerd inject -` the service works great for approximately 2 minutes and then starts to fail with the following errors:

INFO linkerd2_proxy using controller at Some(HostAndPort { host: DnsName(DnsName(DNSName("proxy-api.linkerd.svc.cluster.local"))), port: 8086 })
INFO linkerd2_proxy routing on V4(127.0.0.1:4140)
INFO linkerd2_proxy proxying on V4(0.0.0.0:4143) to None
INFO linkerd2_proxy serving Prometheus metrics on V4(0.0.0.0:4191)
INFO linkerd2_proxy protocol detection disabled for inbound ports {25, 3306}
INFO linkerd2_proxy protocol detection disabled for outbound ports {25, 3306}
INFO admin={bg=resolver} linkerd2_proxy::transport::connect DNS resolved DnsName(DnsName(DNSName("proxy-api.linkerd.svc.cluster.local"))) to 10.0.187.98
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.5.0.42:58096} linkerd2_proxy::proxy::tcp tcp duplex error: Connection reset by peer (os error 104)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.5.0.42:57908} linkerd2_proxy::proxy::tcp tcp duplex error: Connection reset by peer (os error 104)
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.33:38966} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.28:38630} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.28:38630} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.28:38630} linkerd2_proxy::proxy::http::router could not recognize request
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.28:38630} linkerd2_proxy::proxy::http::router could not recognize request

**What you expected to happen**:

The side car proxies continue to work and proxy traffic.

**How to reproduce it (as minimally and precisely as possible)**:

This happens every time I deploy Linkerd with my fairly stock Azure ACS Engine based deployment.

**Anything else we need to know?**:

**Environment**:
- linkerd/namerd version, config files:  linkerd 2.0, installed via `linkerd install | kubectl apply -f -`
- Platform, version, and config files (Kubernetes, DC/OS, etc): Kubernetes version 1.11.3 (server and client)
- Cloud provider or hardware configuration:  Azure ACS Engine

**Comment by timfpark** [comment]
Thanks @olix0r (and sorry for creating the issue in the wrong repo and thanks for moving it to the right one).

Yes, I am seeing errors in the linkerd controller but not the same one:

$ kubectl logs controller-79f5959c45-fp6k5 linkerd-proxy -n linkerd
INFO linkerd2_proxy using controller at Some(HostAndPort { host: DnsName(DnsName(DNSName("localhost."))), port: 8086 })
INFO linkerd2_proxy routing on V4(127.0.0.1:4140)
INFO linkerd2_proxy proxying on V4(0.0.0.0:4143) to None
INFO linkerd2_proxy serving Prometheus metrics on V4(0.0.0.0:4191)
INFO linkerd2_proxy protocol detection disabled for inbound ports {25, 3306}
INFO linkerd2_proxy protocol detection disabled for outbound ports {25, 3306}
INFO admin={bg=resolver} linkerd2_proxy::transport::connect DNS resolved DnsName(DnsName(DNSName("localhost."))) to 127.0.0.1
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.5.0.53:38728} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.5.0.53:38716} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.5.0.53:38736} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.5.0.53:38724} linkerd2_proxy::proxy::tcp tcp duplex error: Connection reset by peer (os error 104)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.5.0.53:38720} linkerd2_proxy::proxy::tcp tcp duplex error: Connection reset by peer (os error 104)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.5.0.53:38768} linkerd2_proxy::proxy::tcp tcp duplex error: Connection reset by peer (os error 104)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.5.0.53:38760} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.5.0.53:38758} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.5.0.53:38766} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.5.0.53:38774} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.5.0.53:38810} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.5.0.53:38808} linkerd2_proxy::proxy::tcp tcp duplex error: Connection reset by peer (os error 104)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.5.0.53:38818} linkerd2_proxy::proxy::tcp tcp duplex error: Connection reset by peer (os error 104)

Here is the other info you requested:

# ldd --version
ldd (Debian GLIBC 2.19-18+deb8u10) 2.19
Copyright (C) 2014 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.

# uname -r
4.15.0-1031-azure

$ linkerd check --pre
kubernetes-api: can initialize the client..................................[ok]
kubernetes-api: can query the Kubernetes API...............................[ok]
kubernetes-api: is running the minimum Kubernetes API version..............[ok]
kubernetes-setup: control plane namespace does not already exist...........[ok]
kubernetes-setup: can create Namespaces....................................[ok]
kubernetes-setup: can create ClusterRoles..................................[ok]
kubernetes-setup: can create ClusterRoleBindings...........................[ok]
kubernetes-setup: can create ServiceAccounts...............................[ok]
kubernetes-setup: can create Services......................................[ok]
kubernetes-setup: can create Deployments...................................[ok]
kubernetes-setup: can create ConfigMaps....................................[ok]
linkerd-version: can determine the latest version..........................[ok]
linkerd-version: cli is up-to-date.........................................[ok]

Status check results are [ok]

$ linkerd check
kubernetes-api: can initialize the client..................................[ok]
kubernetes-api: can query the Kubernetes API...............................[ok]
kubernetes-api: is running the minimum Kubernetes API version..............[ok]
linkerd-api: control plane namespace exists................................[ok]
linkerd-api: control plane pods are ready..................................[ok]
linkerd-api: can initialize the client.....................................[ok]
linkerd-api: can query the control plane API...............................[ok]
linkerd-api[kubernetes]: control plane can talk to Kubernetes..............[ok]
linkerd-api[prometheus]: control plane can talk to Prometheus..............[ok]
linkerd-version: can determine the latest version..........................[ok]
linkerd-version: cli is up-to-date.........................................[ok]
linkerd-version: control plane is up-to-date...............................[ok]

Status check results are [ok]

**Comment by timfpark** [comment]
I just tried this with the latest edge `edge-18.11.2` release and I see the same failure but a slightly different error message in the sidecar proxy to my service:

```
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.4.0.22:44476} linkerd2_proxy::proxy::http::router service error: operation timed out after 10s
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.4.0.22:44478} linkerd2_proxy::proxy::http::router service error: operation timed out after 10s
INFO proxy={server=in listen=0.0.0.0:4143 remote=10.4.0.53:56828} linkerd2_proxy::app::inbound::orig_proto_downgrade downgrading requests; source=Source { remote: V4(10.4.0.53:56828), local: V4(10.4.0.22:4143), orig_dst: Some(V4(10.4.0.22:80)), tls_status: None(Disabled), _p: () }
INFO proxy={server=in listen=0.0.0.0:4143 remote=10.4.0.53:56832} linkerd2_proxy::app::inbound::orig_proto_downgrade downgrading requests; source=Source { remote: V4(10.4.0.53:56832), local: V4(10.4.0.22:4143), orig_dst: Some(V4(10.4.0.22:80)), tls_status: None(Disabled), _p: () }
INFO proxy={server=in listen=0.0.0.0:4143 remote=10.4.0.53:56854} linkerd2_proxy::app::inbound::orig_proto_downgrade downgrading requests; source=Source { remote: V4(10.4.0.53:56854), local: V4(10.4.0.22:4143), orig_dst: Some(V4(10.4.0.22:80)), tls_status: None(Disabled), _p: () }
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.4.0.22:44620} linkerd2_proxy::proxy::http::router service error: operation timed out after 10s
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.4.0.22:44626} linkerd2_proxy::proxy::http::router service error: operation timed out after 10s
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.4.0.22:44668} linkerd2_proxy::proxy::http::router service error: operation timed out after 10s
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.4.0.22:44670} linkerd2_proxy::proxy::http::router service error: operation timed out after 10s
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.4.0.22:44696} linkerd2_proxy::proxy::http::router service error: operation timed out after 10s
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.4.0.22:44714} linkerd2_proxy::proxy::http::router service error: operation timed out after 10s
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.4.0.22:44722} linkerd2_proxy::proxy::http::router service error: operation timed out after 10s
```

I see similar proto_downgrade messages in the controller sidecar, not sure if that is interesting or not:

```
INFO proxy={server=in listen=0.0.0.0:4143 remote=10.4.0.66:55984} linkerd2_proxy::app::inbound::orig_proto_downgrade downgrading requests; source=Source { remote: V4(10.4.0.66:55984), local: V4(10.4.0.77:4143), orig_dst: Some(V4(10.4.0.77:9998)), tls_status: None(Disabled), _p: () }
INFO proxy={server=in listen=0.0.0.0:4143 remote=10.4.0.66:38152} linkerd2_proxy::app::inbound::orig_proto_downgrade downgrading requests; source=Source { remote: V4(10.4.0.66:38152), local: V4(10.4.0.77:4143), orig_dst: Some(V4(10.4.0.77:9996)), tls_status: None(Disabled), _p: () }
INFO proxy={server=in listen=0.0.0.0:4143 remote=10.4.0.66:50110} linkerd2_proxy::app::inbound::orig_proto_downgrade downgrading requests; source=Source { remote: V4(10.4.0.66:50110), local: V4(10.4.0.77:4143), orig_dst: Some(V4(10.4.0.77:9995)), tls_status: None(Disabled), _p: () }
INFO proxy={server=in listen=0.0.0.0:4143 remote=10.4.0.66:56072} linkerd2_proxy::app::inbound::orig_proto_downgrade downgrading requests; source=Source { remote: V4(10.4.0.66:56072), local: V4(10.4.0.77:4143), orig_dst: Some(V4(10.4.0.77:9998)), tls_status: None(Disabled), _p: () }
INFO proxy={server=in listen=0.0.0.0:4143 remote=10.4.0.66:38188} linkerd2_proxy::app::inbound::orig_proto_downgrade downgrading requests; source=Source { remote: V4(10.4.0.66:38188), local: V4(10.4.0.77:4143), orig_dst: Some(V4(10.4.0.77:9996)), tls_status: None(Disabled), _p: () }
INFO proxy={server=in listen=0.0.0.0:4143 remote=10.4.0.66:50144} linkerd2_proxy::app::inbound::orig_proto_downgrade downgrading requests; source=Source { remote: V4(10.4.0.66:50144), local: V4(10.4.0.77:4143), orig_dst: Some(V4(10.4.0.77:9995)), tls_status: None(Disabled), _p: () }
INFO proxy={server=in listen=0.0.0.0:4143 remote=10.4.0.66:56090} linkerd2_proxy::app::inbound::orig_proto_downgrade downgrading requests; source=Source { remote: V4(10.4.0.66:56090), local: V4(10.4.0.77:4143), orig_dst: Some(V4(10.4.0.77:9998)), tls_status: None(Disabled), _p: () }
INFO proxy={server=in listen=0.0.0.0:4143 remote=10.4.0.66:38260} linkerd2_proxy::app::inbound::orig_proto_downgrade downgrading requests; source=Source { remote: V4(10.4.0.66:38260), local: V4(10.4.0.77:4143), orig_dst: Some(V4(10.4.0.77:9996)), tls_status: None(Disabled), _p: () }
INFO proxy={server=in listen=0.0.0.0:4143 remote=10.4.0.66:50218} linkerd2_proxy::app::inbound::orig_proto_downgrade downgrading requests; source=Source { remote: V4(10.4.0.66:50218), local: V4(10.4.0.77:4143), orig_dst: Some(V4(10.4.0.77:9995)), tls_status: None(Disabled), _p: () }
INFO proxy={server=in listen=0.0.0.0:4143 remote=10.4.0.66:56180} linkerd2_proxy::app::inbound::orig_proto_downgrade downgrading requests; source=Source { remote: V4(10.4.0.66:56180), local: V4(10.4.0.77:4143), orig_dst: Some(V4(10.4.0.77:9998)), tls_status: None(Disabled), _p: () }
INFO proxy={server=in listen=0.0.0.0:4143 remote=10.4.0.66:38296} linkerd2_proxy::app::inbound::orig_proto_downgrade downgrading requests; source=Source { remote: V4(10.4.0.66:38296), local: V4(10.4.0.77:4143), orig_dst: Some(V4(10.4.0.77:9996)), tls_status: None(Disabled), _p: () }
INFO proxy={server=in listen=0.0.0.0:4143 remote=10.4.0.66:50252} linkerd2_proxy::app::inbound::orig_proto_downgrade downgrading requests; source=Source { remote: V4(10.4.0.66:50252), local: V4(10.4.0.77:4143), orig_dst: Some(V4(10.4.0.77:9995)), tls_status: None(Disabled), _p: () }
INFO proxy={server=in listen=0.0.0.0:4143 remote=10.4.0.66:56214} linkerd2_proxy::app::inbound::orig_proto_downgrade downgrading requests; source=Source { remote: V4(10.4.0.66:56214), local: V4(10.4.0.77:4143), orig_dst: Some(V4(10.4.0.77:9998)), tls_status: None(Disabled), _p: () }
INFO proxy={server=in listen=0.0.0.0:4143 remote=10.4.0.66:38372} linkerd2_proxy::app::inbound::orig_proto_downgrade downgrading requests; source=Source { remote: V4(10.4.0.66:38372), local: V4(10.4.0.77:4143), orig_dst: Some(V4(10.4.0.77:9996)), tls_status: None(Disabled), _p: () }
INFO proxy={server=in listen=0.0.0.0:4143 remote=10.4.0.66:50344} linkerd2_proxy::app::inbound::orig_proto_downgrade downgrading requests; source=Source { remote: V4(10.4.0.66:50344), local: V4(10.4.0.77:4143), orig_dst: Some(V4(10.4.0.77:9995)), tls_status: None(Disabled), _p: () }
INFO proxy={server=in listen=0.0.0.0:4143 remote=10.4.0.66:56288} linkerd2_proxy::app::inbound::orig_proto_downgrade downgrading requests; source=Source { remote: V4(10.4.0.66:56288), local: V4(10.4.0.77:4143), orig_dst: Some(V4(10.4.0.77:9998)), tls_status: None(Disabled), _p: () }
```

**Comment by olix0r** [comment]
Thanks for the info. I suspect the downgrade messages are benign (this really just means that the proxy is transporting HTTP requests between proxies -- we should lower those to debug message). If you install the control plane or inject a pod and use `--proxy-log-level linkerd2_proxy=trace` you will get a significantly more verbose log that I think might clarify things. If you can share that with me in a gist, that would be helpful.

Additionally, I'm in the process of getting my own azure environment setup (but i'm working through some administrative issues... we'll see how that goes).

**Comment by olix0r** [comment]
I've submitted https://github.com/linkerd/linkerd2-proxy/pull/138 to eliminate that logging noise.

**Comment by timfpark** [comment]
@olix0r As requested, I captured a [trace level log in a gist](https://gist.github.com/timfpark/fd458f451bd998a2a5585043ab481f50) for you.

Thanks again for your help!

**Comment by olix0r** [comment]
@timfpark

```
TRCE proxy={server=in listen=0.0.0.0:4143} linkerd2_proxy::transport::addr_info get_original_dst BoxedIo(TcpStream { addr: V4(10.3.0.93:4143), peer: V4(10.3.0.122:41368), fd: 18 })
...
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.3.0.122:41368} linkerd2_proxy::proxy::http::router could not recognize request
```

This tells us that SO_ORIGINAL_DST is returning `10.3.0.93:4143`, which is the port the proxy is listening on.

Does your application listen on port 4193, by any chance?

It would be helpful to see the log messages from the client (`10.3.0.122`) pod as well.

Thanks for helping us figure this out!

**Comment by timfpark** [comment]
My service should be just listening on port 80 - here is the [pod description](https://gist.github.com/timfpark/30c37ad4d28b4f7b8b51dbd40c819a7b) for it that seems to confirm that.

I've also attached gists of [linkerd-init](https://gist.github.com/timfpark/a007d894982de4b4febd59f1b72907f5), [linkerd-proxy](https://gist.github.com/timfpark/e3c14d0d38028980dc456f8720de6c4a), and the [service itself](https://gist.github.com/timfpark/812cb744501bdaf9b643328099c6841b).

**Comment by dstrebel** [comment]
Having the same issues and using AKS. Let me know if there is anything I can do to help find the root cause.

**Comment by dstrebel** [comment]
@timfpark @olix0r Could this have something to do with api.mapbox.com? I noticed @timfpark in his web request was using api.mapbox.com and I'm actually using the same in my app. Could just be a coincidence though...

**Comment by timfpark** [comment]
If it helps I can provide a kube config for a running cluster with this issue - let me know what would be helpful here or any way I can help

**Comment by hawkw** [comment]
@timfpark Were you able to get the logs from the client proxy that made the request?

**Comment by timfpark** [comment]
@hawkw For me this happens on an incoming request via an Ingress Controller.  Traefik terminates the TLS request and forwards it on to this service as a backend.  So given that, I don't believe there is a Linkerd client - which makes me ask, should I be automatically injecting sidecars for Traefik as well?

**Comment by hawkw** [comment]
@timfpark if you can provide it, I'd also like to see the ingress configuration for the service?

**Comment by timfpark** [comment]
Hi @hawkw - the Ingress is pretty simple:

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    kubernetes.io/ingress.class: traefik
  creationTimestamp: 2018-11-21T22:42:15Z
  generation: 1
  name: map-service-ingress
  namespace: services
  resourceVersion: "3936"
  selfLink: /apis/extensions/v1beta1/namespaces/services/ingresses/map-service-ingress
  uid: b1c22cab-edde-11e8-8cbe-000d3a0d46b9
spec:
  rules:
  - host: map.mydomain.io
    http:
      paths:
      - backend:
          serviceName: map-service
          servicePort: 80
        path: /
status:
  loadBalancer: {}
```

**Comment by timfpark** [comment]
@hawkw - let me know if there is are any other things I can do to help.

@dstrebel - were you seeing this issue on north-south (ingress) traffic only or east-west (service to service) as well?  If the former, what ingress controller were you using?

**Comment by hawkw** [comment]
@timfpark Are you able to get logs from the Traefik ingress controller that's issuing the failing requests? It would be helpful to know what address/host header Traefik is _trying_ to send those requests to.

I have a couple working theories at this point:
1. There's some kind of OS issue that's causing the `getsockopt(..., SO_ORIGINAL_DST, ...)` call to start incorrectly returning the post-iptables-redirection address rather than the original address, so the proxy misidentifies the request target and can't route. This is supported by the fact that both you and @dstrebel are on Azure --- it's possible that there's a specific kernel revision that has this issue that Azure is using as the host OS.
2. There's some kind of Traefik behaviour or Traefik/Linkerd interaction that's  causing Traefik to start sending requests to the proxy's listener address directly, rather than to the app as defined in the ingress spec. I'm not sure what would cause that.

**Comment by hawkw** [comment]
@timfpark I meant to ask about this in the above post, but I'd love to know the _host_ OS's linux kernel version, if you can get that information.

**Comment by olix0r** [comment]
@hawkw OS details are shared here https://github.com/linkerd/linkerd2/issues/1861#issuecomment-438913614

**Comment by hawkw** [comment]
Oops, I missed the `uname -r` output in that comment. Thanks @olix0r.

**Comment by timfpark** [comment]
@hawkw @olix0r Ok - I was able to repro this again today with the latest `edge-18.11.3` release.

What am I seeing now in the linkerd2_proxy logs is that the proxy seems to be failing to resolve api.mapbox.com:

```
ERR! api.mapbox.com:80 linkerd2_proxy::proxy::canonicalize failed to resolve api.mapbox.com: Error(Elapsed)
```

The full linkerd-proxy trace level logs are [here](https://gist.github.com/timfpark/c435b5497973f983587a8ed5189661a8) and the full Traefik debug level logs are [here](https://gist.github.com/timfpark/4072e6d285bde08e81186c2b2a609265).

**Comment by hawkw** [comment]
@timfpark I think the `failed to resolve api.mapbox.com` error seems like issue #1888, which will be fixed in the upcoming edge release.

I believe that error is unrelated to the `could not recognize request` error, which in this case is being logged by the proxy's _inbound_ router --- it fails to recognize a request that was sent from Traefik to the proxy. Meanwhile, the request from the service behind the proxy to `api.mapbox.com` would be handled by the proxy's _outbound_ router, so if that was causing the `could not recognize request` error, we'd see that being logged by the outbound proxy.

**Comment by hawkw** [comment]
Thanks for the Traefik logs, by the way --- I'm taking a look at them now.

**Comment by hawkw** [comment]
@timfpark searching the Traefik logs you've posted in https://github.com/linkerd/linkerd2/issues/1861#issuecomment-444665100, I can't find any mention of the IP `10.0.6.169` (the `map-service` cluster-IP according to https://github.com/linkerd/linkerd2/issues/1861#issuecomment-443810638), nor of `map.rhom.io`, or of `10.5.0.36:4143`, the inbound listener address for the proxy in the proxy logs you posted.

Furthermore, in the proxy logs, it seems that inbound requests to `map.rhom.io` are being routed correctly. It's possible that the requests sent directly to the proxy's inbound listener (`10.5.0.36:4143`) are coming from some other source?

Are you sure the Traefik instance that generated those logs is actually talking to the proxy?

**Comment by timfpark** [comment]
I believe there are only two sources of traffic talking to the map-service:

1. Prometheus polling it on /metrics internally from within the cluster.
2. Traefik ingressing traffic to it from outside the cluster.

Just in case I made I mistake with the logs I included before and mismatched two different cluster's IP addresses, here is a full fresh set again of [traefik ingress](https://gist.github.com/timfpark/fe6fbc15e767b94c1427e004c9152559), [linkerd-proxy trace level logs](https://gist.github.com/timfpark/8ba10d4a6f9cd2310e0f757344f587d3), and a [pod-ip mapping](https://gist.github.com/timfpark/8bfeba1d93d425c57f329fedca45eab7).

**Comment by hawkw** [comment]
@timfpark Ah, okay, thanks for the pod-ip mapping! Looking at the linkerd logs, it looks like the requests that couldn't be recognized are issuing from the IP `10.5.0.125`:
```
ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.5.0.125:58934} linkerd2_proxy::proxy::http::router could not recognize request
```

According to the list of pods in your gist, `10.5.0.125` corresponds to your Prometheus instance:
```
prometheus prometheus-server-567b58bbd7-5w6hk 2/2 Running 0 6h 10.5.0.125 k8s-agents-18763741-1 <none>
```

It looks like Prometheus is trying to scrape the proxy's inbound listener IP, and is issuing requests to that IP directly. The proxy expects requests sent to the inbound listener to be targeted at the proxied `map-service` instead, so it fails to route the scrapes.

Note that linkerd's proxy serves its own Prometheus metrics on port `4191`, so if you want to collect the proxy's metrics, you should have Prometheus scrape that port instead. Alternatively, the Linkerd control plane runs its own Prometheus instance in the `linkerd` namespace, which should also have the metrics from all Linkerd proxies, so you could use [Prometheus federation](https://prometheus.io/docs/prometheus/latest/federation/) instead.

**Comment by timfpark** [comment]
Ok - so if I am understanding correctly, it sounds like there are two separate things happening here:

1. There is a DNS resolution failure of `api.mapbox.com` that is potentially causing the end problem I am seeing and is possibly fixed in the next edge release #1888.
2. I have a misconfiguration of Prometheus in my cluster that is causing the original problem.  One thing I wasn't clear on in your response is:  how do I scrape metrics from the map-service itself?  Will the proxy surface those metrics as well?  (Sorry if this is covered somewhere in the docs that I am missing.) 

Thanks for your help!

**Comment by hawkw** [comment]
@timfpark Yeah, that's correct. 

To scrape metrics from the map service itself, your Prometheus should be set to scrape whatever port the map service is serving metrics on. Linkerd's iptables rules will automatically redirect that to the proxy's inbound router, and the proxy will detect the original destination port and forward the request to the service, just like it does for all other incoming traffic.

**Comment by hawkw** [comment]
I'm closing this issue, as I think all the questions here have been addressed. @timfpark, if you have any more trouble, please feel free to re-open this, or open a new issue.

**Comment by timfpark** [comment]
Yes - confirmed fixed with `stable-2.1.0` - thanks again for all your help!

**Comment by hawkw** [comment]
Great, I'm glad to hear everything's working now!
